pkg/protocol/tipmanager/v1/tipmanager.go

```diff
@@ -0,0 +1,363 @@
+
+		return true
+	})
+
+	return tipSet
+}
+
+// WeakTipSet returns the weak tip set of the TipManager.
+func (t *TipManager) WeakTipSet() (tipSet []*blocks.Block) {
+	t.weakTipSet.ForEach(func(_ iotago.BlockID, tipMetadata *TipMetadata) bool {
+		tipSet = append(tipSet, tipMetadata.Block())
+
+		return true
+	})
+
+	return tipSet
+}
+
+// Evict evicts a slot from the TipManager.
+func (t *TipManager) Evict(slotIndex iotago.SlotIndex) {
+	if t.markSlotAsEvicted(slotIndex) {
+		if evictedObjects, deleted := t.tipMetadataStorage.DeleteAndReturn(slotIndex); deleted {
+			evictedObjects.ForEach(func(_ iotago.BlockID, tipMetadata *TipMetadata) bool {
+				tipMetadata.evicted.Trigger()
+
+				return true
+			})
+		}
+	}
+}
+
+// Events returns the events of the TipManager.
+func (t *TipManager) Events() *tipmanager.Events {
+	return t.events
+}
+
+// Shutdown marks the TipManager as shutdown.
+func (t *TipManager) Shutdown() {
+	t.TriggerStopped()
+}
+
+// setupBlockMetadata sets up the behavior of the given Block.
+func (t *TipManager) setupBlockMetadata(tipMetadata *TipMetadata) {
+	tipMetadata.stronglyConnectedToTips.OnUpdate(func(_, isConnected bool) {
+		t.updateParents(tipMetadata, propagateConnectedChildren(isConnected, true))
+	})
+
+	tipMetadata.weaklyConnectedToTips.OnUpdate(func(_, isConnected bool) {
+		t.updateParents(tipMetadata, propagateConnectedChildren(isConnected, false))
+	})
+
+	tipMetadata.OnIsStrongTipUpdated(func(isStrongTip bool) {
+		if isStrongTip {
+			t.strongTipSet.Set(tipMetadata.Block().ID(), tipMetadata)
+		} else {
+			t.strongTipSet.Delete(tipMetadata.Block().ID())
+		}
+	})
+
+	tipMetadata.OnIsWeakTipUpdated(func(isWeakTip bool) {
+		if isWeakTip {
+			t.weakTipSet.Set(tipMetadata.Block().ID(), tipMetadata)
+		} else {
+			t.weakTipSet.Delete(tipMetadata.Block().ID())
+		}
+	})
+
+	tipMetadata.OnEvicted(func() {
+		t.strongTipSet.Delete(tipMetadata.Block().ID())
+		t.weakTipSet.Delete(tipMetadata.Block().ID())
+	})
+
+	tipMetadata.setTipPool(t.determineTipPool(tipMetadata))
+
+	t.events.BlockAdded.Trigger(tipMetadata)
+}
+
+// determineTipPool determines the initial TipPool of the given Block.
+func (t *TipManager) determineTipPool(tipMetadata *TipMetadata, minPool ...tipmanager.TipPool) tipmanager.TipPool {
+	if lo.First(minPool) <= tipmanager.StrongTipPool && !t.conflictDAG.AcceptanceState(tipMetadata.Block().ConflictIDs()).IsRejected() {
+		return tipmanager.StrongTipPool
+	}
+
+	if lo.First(minPool) <= tipmanager.WeakTipPool && t.conflictDAG.LikedInstead(tipMetadata.Block().PayloadConflictIDs()).Size() == 0 {
+		return tipmanager.WeakTipPool
+	}
+
+	return tipmanager.DroppedTipPool
+}
+
+// updateParents updates the parents of the given Block.
+func (t *TipManager) updateParents(tipMetadata *TipMetadata, updates map[model.ParentsType]func(*TipMetadata)) {
+	if parentBlock := tipMetadata.Block(); parentBlock != nil && parentBlock.Block() != nil {
+		parentBlock.ForEachParent(func(parent model.Parent) {
```

Comment:
This `ForEachParent` is imo a potential hazard. It skips duplicate parents ie if a parent is a strong and a like parent it is only in there once as a strong parent. Not sure if that is an issue here but just double checking. I think we should adjust the way `ForEachParent` works.

Reply:
This indeed sounds very problematic, I wasn't aware of this "quirk".

I agree that we should most probably change the ForEachParent to return everything. Is there any method that would "expect" the method to magically drop parents?

Reply:
I switched to use ParentsWithType